### PR TITLE
Enable image and model downloading

### PR DESCRIPTION
**Pull request by amickan**
This enables serving algorithm images and models to users with download permission. 

Part of https://github.com/DIAGNijmegen/rse-roadmap/issues/310 

This also requires some minor changes to our terraform repo, a cloudfront distribution for the private S3 bucket and the following environment variables:
`PRIVATE_S3_STORAGE_CLOUDFRONT_DOMAIN `
`PRIVATE_S3_STORAGE_USE_CLOUDFRONT`

### CODE REVIEW

File: app/grandchallenge/algorithms/models.py [hunk 1]
```diff
@@ -567,6 +570,13 @@ def get_absolute_url(self):
             kwargs={"slug": self.algorithm.slug, "pk": self.pk},
         )
 
+    @property
+    def download_url(self):
```
Comment (amickan) [inline app/grandchallenge/algorithms/models.py:574]: This link (and the corresponding one for the model below) are not used yet, but will be once the remaining external evaluation changes are implemented, hence I added them already.

File: app/config/settings.py [hunk 2]
```diff
@@ -190,7 +190,13 @@
 PRIVATE_S3_STORAGE_KWARGS = {
     "bucket_name": os.environ.get(
         "PRIVATE_S3_STORAGE_BUCKET_NAME", "grand-challenge-private"
-    )
+    ),
+    # This is the domain where people will be able to go to download data
+    # from this bucket. Usually we would use reverse to find this out,
+    # but this needs to be defined before the database is populated
```
Comment (jmsmkn) [inline app/config/settings.py:196]: Can't you use `reverse_lazy`?

File: app/config/settings.py [hunk 3]
```diff
@@ -210,6 +216,12 @@
 PROTECTED_S3_STORAGE_CLOUDFRONT_DOMAIN = os.environ.get(
     "PROTECTED_S3_STORAGE_CLOUDFRONT_DOMAIN_NAME", ""
 )
+PRIVATE_S3_STORAGE_CLOUDFRONT_DOMAIN = os.environ.get(
+    "PRIVATE_S3_STORAGE_CLOUDFRONT_DOMAIN_NAME", ""
+)
+PRIVATE_S3_STORAGE_USE_CLOUDFRONT = strtobool(
+    os.environ.get("PRIVATE_S3_STORAGE_USE_CLOUDFRONT", "False")
+)
```
Comment (jmsmkn) [inline app/config/settings.py:224]: We won't want to use CloudFront for the private storage.

**Comment by jmsmkn** [comment]
I think the approach here isn't quite right. We will not want to add CloudFront to the private storage. All that needs to happen is that a signed S3 url is generated to download the object. I'll see if I can migrate the algorithm images to the protected bucket quickly.

**Comment by amickan** [comment]
> I think the approach here isn't quite right. We will not want to add CloudFront to the private storage. All that needs to happen is that a signed S3 url is generated to download the object. I'll see if I can migrate the algorithm images to the protected bucket quickly so this might not be needed at all.

Alright. I was unsure about using CloudFront. Could you elaborate why we shouldn't use CloudFront for this use case?

**Comment by amickan** [comment]
> Could you elaborate why we shouldn't use CloudFront for this use case?

Ah, I misunderstood your comment and though we didn't want to use CloudFront for algorithm image downloading, but you mean for the private bucket in general. That does make sense actually.

**Comment by amickan** [comment]
> > Could you elaborate why we shouldn't use CloudFront for this use case?
> 
> Ah, I misunderstood your comment and thought you meant that we didn't want to use CloudFront for algorithm image downloading, but you mean for the private bucket in general. That does make sense actually.

**Comment by amickan** [comment]
Closed in favour of https://github.com/comic/grand-challenge.org/pull/3409

**Pull request closed without merging**